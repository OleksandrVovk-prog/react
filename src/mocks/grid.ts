import { getRandomColor } from '../utils/helpers';

export default {
  squares: [
    {
      id: Date.now(),
      x: 0,
      y: 0,
      color: getRandomColor(),
    },
  ],
  itemName: 'Eddie',
};
