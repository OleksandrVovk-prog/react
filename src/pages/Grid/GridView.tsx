import { type ReactElement, useEffect } from 'react';
import Draggable from 'react-draggable';
import { toast, ToastContainer } from 'react-toastify';

import { GRID_HEIGHT, GRID_WIDTH } from '../../constants/Grid';
import { getContrastColor } from '../../utils/helpers';

import type { IGridView } from './interfaces/IGridView';

import styles from './sass/GridView.module.scss';

function GridView({
  squares,
  contextMenu,
  onHandleRightClick,
  onHandleCreateSquare,
  onHideContextMenu,
  errorMessage,
  setErrorMessage,
  handleDragStop,
  handleDrag,
}: IGridView): ReactElement {
  useEffect(() => {
    if (errorMessage) {
      toast.error(errorMessage, {
        onClose: () => {
          setErrorMessage('');
        },
      });
    }
  }, [errorMessage]);

  return (
    <div
      aria-hidden
      className={styles.gridContainer}
      onClick={onHideContextMenu}
      onContextMenu={onHandleRightClick}
      style={{ backgroundSize: `${GRID_WIDTH}px ${GRID_HEIGHT}px` }}
    >
      <ToastContainer
        position="top-right"
        autoClose={1000}
        hideProgressBar
        newestOnTop={false}
        closeOnClick
        rtl={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover
      />
      {contextMenu ? (
        <div
          className={styles.gridContextMenu}
          style={{
            top: contextMenu.y,
            left: contextMenu.x,
          }}
        >
          <div aria-hidden onClick={onHandleCreateSquare}>
            Create Square
          </div>
        </div>
      ) : null}
      {squares?.map((square) => (
        <Draggable
          key={square.id}
          defaultPosition={{ x: square.x, y: square.y }}
          position={{ x: square.x, y: square.y }}
          grid={[GRID_WIDTH, GRID_HEIGHT]}
          onStop={(e, data) => handleDragStop(square.id, e, data)}
          onDrag={(e, data) => handleDrag(square.id, e, data)}
        >
          <div
            className={styles.gridItem}
            style={{
              width: GRID_WIDTH,
              height: GRID_HEIGHT,
              backgroundColor: square.color,
              color: getContrastColor(square.color),
            }}
          >
            Eddie
          </div>
        </Draggable>
      ))}
    </div>
  );
}

export default GridView;
