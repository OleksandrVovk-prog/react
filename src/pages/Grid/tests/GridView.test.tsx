import { screen } from '@testing-library/react';

import { initServer } from '../../../../__tests__/mswServer';
import { render } from '../../../../__tests__/test-utils';
import grid from '../../../mocks/grid';
import GridView from '../GridView';

describe('<GridView />', () => {
  initServer();
  test('should render component', async () => {
    const { asFragment } = render(
      <GridView
        squares={grid.squares}
        contextMenu={null}
        onHandleRightClick={() => {}}
        onHandleCreateSquare={() => {}}
        onHideContextMenu={() => {}}
        errorMessage=""
        setErrorMessage={() => {}}
        handleDragStop={() => {}}
        handleDrag={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('should render content after loading', async () => {
    render(
      <GridView
        squares={grid.squares}
        contextMenu={null}
        onHandleRightClick={() => {}}
        onHandleCreateSquare={() => {}}
        onHideContextMenu={() => {}}
        errorMessage=""
        setErrorMessage={() => {}}
        handleDragStop={() => {}}
        handleDrag={() => {}}
      />,
    );
    expect(screen.getByText(grid.itemName)).toBeInTheDocument();
  });
});
