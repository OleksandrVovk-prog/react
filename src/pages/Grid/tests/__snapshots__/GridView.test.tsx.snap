// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridView /> should render component 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="gridContainer"
    style="background-size: 50px 50px;"
  >
    <section
      aria-atomic="false"
      aria-label="Notifications Alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      class="Toastify"
    />
    <div
      class="gridItem react-draggable"
      style="width: 50px; height: 50px; background-color: rgb(34, 178, 158); color: rgb(0, 0, 0); transform: translate(0px,0px);"
    >
      Eddie
    </div>
  </div>
</DocumentFragment>
`;
